src/types.jl

```diff
@@ -1,9 +1,33 @@
 """
-`Compiled` is a trait indicating that any `:call` expressions should be evaluated
-using Julia's normal compiled-code evaluation. The alternative is to pass `stack=Frame[]`,
-which will cause all calls to be evaluated via the interpreter.
+    abstract type Interpreter end
+
+An interpreter that subtypes this type can implement its own evaluation strategies, by
+overloading the certain methods in JuliaInterpreter that are defined for this base type.
+The default behaviors of `Interpreter` is same as that of [`RecursiveInterpreter`](@ref),
+meaning it will recursively interpret all `:call` expressions.
+"""
+abstract type Interpreter end
+
 """
-struct Compiled end
+    RecursiveInterpreter <: Interpreter
+
+`RecursiveInterpreter` is an [`Interpreter`](@ref) that recursively interprets any `:call`
+expressions in the code being interpreted.
+"""
+struct RecursiveInterpreter <: Interpreter end
+
+"""
+    NonRecursiveInterpreter <: Interpreter
+
+`NonRecursiveInterpreter` is an [`Interpreter`](@ref) that evaluates any `:call` expressions
+in the code being interpreted using Julia's normal code execution engine with the native
+compiler.
+
+`JuliaInterpreter.Compiled` is alised to `NonRecursiveInterpreter` for backward compatibility.
+"""
+struct NonRecursiveInterpreter <: Interpreter end
+
+const Compiled = NonRecursiveInterpreter # for backward compatibility
 Base.similar(::Compiled, sz) = Compiled()  # to support similar(stack, 0)
```

Comment:
```suggestion
Base.similar(::NonRecursiveInterpreter, sz) = NonRecursiveInterpreter()  # to support similar(stack, 0)
```

Reply:
This method is required for the `Compiled` usages where it is used as a special token indicating somethings need to be compiled, so I'd like to leave this method definition as is for now.